lib/temporal/workflow/dispatcher.rb

```diff
@@ -6,21 +6,33 @@ class Dispatcher
 
       def initialize
         @handlers = Hash.new { |hash, key| hash[key] = [] }
+        @query_handlers = Hash.new { |hash, key| hash[key] = {} }
```

Comment:
I would recommend keeping the `Dispatcher` unaware of specific use-cases (such as queries). It's a general event dispatcher and there's really no difference between queries and other events. We can use the `target` to identify a specific query

Reply:
Can you elaborate on using the `target` to identify the specific query? I don't think I fully understand any potential implications of using a different `target`. Also if there's a design that would accommodate specific signal handler registration as well?

Reply:
Take a look at the context.rb for how the signals are registered and dispatched. Queries should be exactly the same. They have a target, a name, and a block which is identical to signals. So instead of tracking them in their own `query_handlers` hash, use the main one.

In context.rb, you'll register like this:
```
dispatcher.register_handler(target, query) { |input| block.call(input) }
```

To call it in executor.rb, use `dispatch` directly just like we did for signals.
```
dispatcher.dispatch(History::EventTarget.workflow, query, args)
```
The code to handle the special-case query `__cadence_web_list` would be either in task processor or executor. Don't leak it into here.

Reply:
Hmmm, the query may conflict with event_name? I honestly don't understand this. Is there a simple example to illustrate? Otherwise, the query_handlers and handlers structures are doing the same thing as far as I can tell. You look it up by target and then match on query/event_name.

Reply:
(As far as I understand) if you were to register a `workflow.on_query('started')` it would add a handler function in `handlers[target]` with a structure of `['started', handler]`. This event name would also match and have its handler invoked for internally dispatched 'started' events that share the same target. If we want to share the data structure (with the same target lookup), something in the `event_name` would have to differentiate it as a query handler.

Reply:
Okay, I understand the issue now. Thanks for spelling it out.

My suggestion is we solve this by adopting a convention. `register_handler` is called internally where there could be conflicts in the future ('failed', 'signaled', 'completed'). Internal usages of `register_handler` should adopt a convention of two leading underscores for the event_name and then we document that for the SDK so that end users don't do the same. 

There could be a potential conflict with signal names (if we enhance `on_signal` to take a name like you did with `on_query`). It's up to the programmer to avoid name collisions but our docs could suggest a convention there where signal names start with `s_` and query names start with `q_`.

Think that gets us the result we want?